impeller/renderer/backend/vulkan/graphics_queue_vk.cc

```diff
@@ -0,0 +1,91 @@
+#include "impeller/base/validation.h"
+#include "impeller/renderer/backend/vulkan/command_buffer_vk.h"
+#include "impeller/renderer/backend/vulkan/command_encoder_vk.h"
+#include "impeller/renderer/backend/vulkan/context_vk.h"
+#include "impeller/renderer/backend/vulkan/fence_waiter_vk.h"
+#include "impeller/renderer/backend/vulkan/graphics_queue_vk.h"
+#include "impeller/renderer/backend/vulkan/tracked_objects_vk.h"
+#include "impeller/renderer/command_buffer.h"
+
+namespace impeller {
+
+GraphicsQueueVK::GraphicsQueueVK(const std::weak_ptr<ContextVK>& context)
+    : context_(context) {}
+
+fml::Status GraphicsQueueVK::Submit(
+    const std::vector<std::shared_ptr<CommandBuffer>>& buffers,
+    const CompletionCallback& callback) {
+  if (buffers.empty()) {
+    return fml::Status(fml::StatusCode::kInvalidArgument,
+                       "No command buffers provided.");
+  }
+  bool fail_callback = !!callback;
+  // Success or failure, you only get to submit once.
+  fml::ScopedCleanupClosure reset([&]() {
+    if (fail_callback) {
+      callback(CommandBuffer::Status::kError);
+    }
+  });
+
+  std::vector<vk::CommandBuffer> vk_buffers;
+  std::vector<std::shared_ptr<TrackedObjectsVK>> tracked_objects;
+  vk_buffers.reserve(buffers.size());
+  tracked_objects.reserve(buffers.size());
+  for (const std::shared_ptr<CommandBuffer>& buffer : buffers) {
+    auto encoder = CommandBufferVK::Cast(*buffer).GetEncoder();
+    if (!encoder->EndCommandBuffer()) {
+      return fml::Status(fml::StatusCode::kCancelled,
+                         "Failed to end command buffer.");
+    }
+    tracked_objects.push_back(encoder->tracked_objects_);
+    vk_buffers.push_back(encoder->GetCommandBuffer());
+    encoder->Reset();
+  }
+
+  auto context = context_.lock();
+  if (!context) {
+    VALIDATION_LOG << "Device lost.";
+    return fml::Status(fml::StatusCode::kCancelled, "Device lost.");
+  }
+  auto [fence_result, fence] = context->GetDevice().createFenceUnique({});
+  if (fence_result != vk::Result::eSuccess) {
+    VALIDATION_LOG << "Failed to create fence: " << vk::to_string(fence_result);
+    return fml::Status(fml::StatusCode::kCancelled, "Failed to create fence.");
+  }
+
+  vk::SubmitInfo submit_info;
+  submit_info.setCommandBuffers(vk_buffers);
+  auto status = context->GetGraphicsQueue()->Submit(submit_info, *fence);
+  if (status != vk::Result::eSuccess) {
+    VALIDATION_LOG << "Failed to submit queue: " << vk::to_string(status);
+    return fml::Status(fml::StatusCode::kCancelled, "Failed to submit queue: ");
+  }
+
+  // Submit will proceed, call callback with true when it is done and do not
+  // call when `reset` is collected.
+  auto added_fence = context->GetFenceWaiter()->AddFence(
+      std::move(fence),
+      [callback, tracked_objects = std::move(tracked_objects)]() mutable {
+        // Ensure tracked objects are destructed before calling any final
+        // callbacks.
+        for (auto& tracked_object : tracked_objects) {
```

Comment:
`tracked_objects.clear()` is the same as individually resetting the items in a loop.